peekaboo/toolbox/cortex.py

```diff
@@ -0,0 +1,632 @@
+    def __init__(self, analyzer):
+        super().__init__("Cortex analysis report for analyzer '%s' is missing"
+                         % analyzer)
+        self.analyzer = analyzer
+
+
+class CortexAnalyzerReport:
+    """ Cortex analyzer report base class. """
+
+
+class CortexAnalyzer:
+    """ Cortex analyzer base class. """
+    name = 'unknown'
+    reportclass = CortexAnalyzerReport
+
+
+class CortexFileAnalyzer:
+    """ An analyzer which accepts a file as main input. """
+    def get_submit_parameters(self, sample, submit_original_filename=False):
+        """ Return this analyzer's submit parameters for a given sample. """
+        path = sample.submit_path
+        filename = os.path.basename(path)
+
+        # override with the original file name if available
+        if submit_original_filename:
+            if sample.name_declared:
+                filename = sample.name_declared
+            elif sample.filename:
+                filename = sample.filename
+
+        params = {
+            'data': path,
+            'dataType': 'file',
+            'tlp': 1,
+            'parameters': {
+                'filename': filename,
+            }
+        }
+
+        return params
+
+
+class CortexHashAnalyzer(CortexAnalyzer):
+    """ An analyzer which accepts hashes as main input. """
+    def get_submit_parameters(self, sample, submit_original_filename=False):
+        """ Return this analyzer's submit parameters for a given sample. """
+        params = {
+            'data': sample.sha256sum,
+            'dataType': 'hash',
+            'tlp': 1,
+        }
+
+        return params
+
+
+class FileInfoAnalyzerReport(CortexAnalyzerReport):
+    """ Represents a Cortex FileInfo_6_0 analysis JSON report. """
+    def __init__(self, report):
+        self.report = report
+
+    @property
+    def full(self):
+        """ Return the full report. """
+        return self.report.get('full', None)
+
+
+class FileInfoAnalyzer(CortexFileAnalyzer):
+    """ Interfaces with Cortex Analyzer FileInfo_6_0. """
+    name = 'FileInfo_6_0'
+    reportclass = FileInfoAnalyzerReport
+
+
+class HybridAnalysisReport(CortexAnalyzerReport):
+    """ Represents a Cortex HybridAnalysis_GetReport_1_0 analysis JSON
+        report. """
+    def __init__(self, report):
+        self.report = report
+
+    @property
+    def full(self):
+        """ Return the full report. """
+        return self.report.get('full', None)
+
+
+class HybridAnalysis(CortexFileAnalyzer):
+    """ Interfaces with Cortex Analyzer HybridAnalysis_GetReport_1_0. """
+    name = 'HybridAnalysis_GetReport_1_0'
+    reportclass = HybridAnalysisReport
+
+
+class VirusTotalQueryReport(CortexAnalyzerReport):
+    """ Represents a Cortex VirusTotal_GetReport_3_0 analysis JSON report. """
+    def __init__(self, report):
+        self.report = report
+        self.taxonomies = report.get("summary", {}).get("taxonomies", [{}])[0]
```

Comment:
Have a look at the current code, most recent commit. I dropped the `print()` because it seemed debug-ish. ;)

Also, I streamlined the use of `self.taxonomies` because in the case where we don't extract a dict in the loop we'd have ended up with a list of an empty dict (`[{}]`) which might have caused fallout. Does the adjusted code make sense?

Can you give me a reason why we're doing this exactly for the commit message?

Also, I just realise, we'd take the last `tax` with namespace `VT`, not the first. Is there only every one or should we accumulate or merge them or should we break at the first hit?

Reply:
Any update on the rationale of this change and correctness of my code changes?

Reply:
This is a report by Cortex  VirusTotal_GetReport_3_0 
```
{
  "summary": {
    "taxonomies": [
      {
        "level": "malicious",
        "namespace": "VT",
        "predicate": "GetReport",
        "value": "37/68"
      }
    ]
  },
  "full": {
    "scans": {
      "Bkav": {
....
```

I have changes to the respective code part, will push, and await your comment

Reply:
I'll try to answer my own questions from my reading of the code:
* the `print()` was a debug remnant.
* the new helper function enforces a constraint that there is only ever one taxonomy of namespace `VT`
* due to that constraint, the question of picking the first or last list element becomes moot
* the rationale for the change is input validation: We do not want to just accept anything but explicitly look for the expected namespace and make sure there's only one entry.

Correct?

If so, can you tune the messages of both commits that effect, maybe even squash them together?